Building your Own Hooks Translation (#34)
reactjs/es.reactjs.org: content/docs/hooks-custom.md

---
id: hooks-custom
title: Construyendo Tus Propios Hooks
permalink: docs/hooks-custom.html
next: hooks-reference.html
prev: hooks-rules.html
---
*Hooks* are an upcoming feature that lets you use state and other React features without writing a class. They're currently in React v16.8.0-alpha.1.

Construir tus propios Hooks te permite extraer la lógica del componente en funciones reutilizables.

> Cuando estábamos aprendiendo a [usar el Efecto de Hook](/docs/hooks-effect.html#example-using-hooks-1), vimos este componente de una aplicación de chat que muestra un mensaje indicando si un amigo está conectado o desconectado:

Reviewer: I think it should be used "El Hook de efecto", "El Hook de estado", etc.
```suggestion
Cuando estábamos aprendiendo a [usar el Hook de efecto](/docs/hooks-effect.html#example-using-hooks-1), vimos este componente de una aplicación de chat que muestra un mensaje indicando si un amigo está conectado o desconectado:
```

}
```

> Ahora vamos a decir que nuestra aplicación de chat tiene también una lista de contactos y queremos que renderice nombres de usuarios con color verde. Podríamos copiar y pegar lógica similar sobre nuestro componente `FriendListItem` pero eso no seria ideal:

Reviewer: I think paraphrasing here helps to translate better the meaning.
```suggestion
Supongamos ahora que nuestra aplicación de chat tiene también una lista de contactos y queremos que renderice nombres de usuarios con color verde. Podríamos copiar y pegar la lógica anterior adaptada a nuestro componente `FriendListItem`, pero eso no sería ideal:
```

}
```

En cambio, nos gustaría compartir esta lógica entre `FriendStatus` y `FriendListItem`.

Tradicionalmente en React, hemos tenido dos formas populares para compartir lógica de estados entre componentes: [renderizar props](/docs/render-props.html) y [componentes de orden mas alto](/docs/higher-order-components.html). Ahora veremos como los Hooks resuelven muchos de los mismos problemas sin forzarte a añadir más componentes al árbol.

## Extrayendo un Hook Personalizado

Cuando queremos compartir lógica entre dos funciones de Javascript, lo extraemos en una tercera función. Ambos, componentes y Hooks, son funciones, así que esto funciona para ellos también!

> **Un Hook personalizado es una función de JavaScript cuyo nombre comienza con "`use`" y que puede llamar a otros Hooks.** Por ejemplo, `useFriendStatus` abajo es nuestro primer Hook personalizado:

Reviewer: I would allways use "anterior" or "a continuación". This is a similar pattern for the English "above", "below"
```suggestion
**Un Hook personalizado es una función de JavaScript cuyo nombre comienza con "`use`" y que puede llamar a otros Hooks.** Por ejemplo, a continuación, `useFriendStatus` es nuestro primer Hook personalizado:
```

}
```

> No hay nada nuevo dentro de este -- la lógica es copiada de los componentes anteriores. Al igual que en un componente, asegúrate de solo llamar a otros Hooks incondicionalmente en el nivel superior de tu Hook personalizado.

Reviewer: The use of the em-dash here is an orthotypographical Anglicism. 
```suggestion
No hay nada nuevo dentro, la lógica se copia de los componentes anteriores. Al igual que en un componente, asegúrate de solo llamar a otros Hooks incondicionalmente en el nivel superior de tu Hook personalizado.
```

A diferencia de un componente de React, un Hook personalizado no necesita tener una firma específica. Podemos decidir lo que adopta como argumentos y que, si lo hace, debería devolver. En otras palabras, es como una función normal. Su nombre debería siempre empezar con `use` así se puede decir que de un vistazo las [reglas de Hooks](/docs/hooks-rules.html) se le aplican.

> El propósito de nuestro Hook `useFriendStatus` es suscribirnos al estado de un amigo. Por esto toma a `friendID` como un argumento, y devuelve cuando este amigo esta conectado:

Reviewer: It does not return *when* he is online, but *whether* is online
```suggestion
El propósito de nuestro Hook `useFriendStatus` es suscribirnos al estado de un amigo. Por esto toma a `friendID` como un argumento, y devuelve si este amigo está conectado:
```

}
```

Ahora veamos como podemos usar nuestro Hook personalizado.

## Usando un Hook Personalizado

> Al principio, nuestro objetivo declarado fue eliminar la lógica duplicada de los componentes `FriendStatus` y `FriendListItem`. Ambos quieren saber cuando tu amigo esta conectado.

Reviewer: ```suggestion
Al principio, nuestro objetivo declarado fue eliminar la lógica duplicada de los componentes `FriendStatus` y `FriendListItem`. Ambos quieren saber cuando un amigo está conectado.
```

}
```

> **Este código es equivalente a los ejemplos originales?** Si, funciona exactamente de la misma forma. Si miras de cerca, notarás que no hicimos cambios en el comportamiento. Todo lo que hicimos fue extraer código común entre dos funciones en una función separada. **Los Hooks personalizados son una convención que naturalmente sigue el diseño de los Hooks, en lugar de una característica de React.**

Reviewer: ```suggestion
**¿Este código es equivalente a los ejemplos originales?** Sí, funciona exactamente de la misma forma. Si miras de cerca, notarás que no hicimos cambios en el comportamiento. Todo lo que hicimos fue extraer código común entre dos funciones en una función separada. **Los Hooks personalizados son una convención que surge naturalmente del diseño de los Hooks, en lugar de una característica de React.**
```

> **Tengo que nombrar mis Hooks personalizados comenzando con "`use`"?** Por favor hazlo. Esta convención es muy importante. Sin esta, no podríamos comprobar automáticamente violaciones de [ las reglas de los Hooks](/docs/hooks-rules.html) porque no podríamos decir si una cierta función contiene llamados a Hooks dentro de la misma.

Reviewer: ```suggestion
**¿Tengo que nombrar mis Hooks personalizados comenzando con "`use`"?** Por favor, hazlo. Esta convención es muy importante. Sin esta, no podríamos comprobar automáticamente violaciones de [las reglas de los Hooks](/docs/hooks-rules.html) porque no podríamos decir si una cierta función contiene llamados a Hooks dentro de la misma.
```

> **Dos componentes usando el mismo Hook comparten estado?** No. Los Hooks personalizados son un mecanismo para reutilizar *lógica de estado * (como configurar una suscripción y recordar el valor actual), pero cada vez que usas un Hook personalizado, todo estado y efecto dentro de este son aislados completamente.

Reviewer: ```suggestion
**¿Dos componentes usando el mismo Hook comparten estado?** No. Los Hooks personalizados son un mecanismo para reutilizar *lógica de estado * (como configurar una suscripción y recordar el valor actual), pero cada vez que usas un Hook personalizado, todo estado y efecto dentro de este son aislados completamente.
```

> **Cómo un Hook personalizado obtiene un estado aislado?** Cada *llamada* al Hook obtiene un estado aislado. Porque llamamos `useFriendStatus` directamente, desde el punto de vista de React nuestro componente llama a `useState` y `useEffect`. Y como [ aprendimos](/docs/hooks-effect.html#tip-use-multiple-effects-to-separate-concerns) [anteriormente](/docs/hooks-effect.html#tip-use-multiple-effects-to-separate-concerns) podemos llamar a `useState` y a ` useEffect` muchas veces en un momento y ellos van a ser completamente independientes.

Reviewer: ```suggestion
*¿*Cómo un Hook personalizado obtiene un estado aislado?** Cada *llamada* al Hook obtiene un estado aislado. Porque llamamos `useFriendStatus` directamente, desde el punto de vista de React nuestro componente llama a `useState` y `useEffect`. Y como [aprendimos](/docs/hooks-effect.html#tip-use-multiple-effects-to-separate-concerns) [anteriormente](/docs/hooks-effect.html#tip-use-multiple-effects-to-separate-concerns) podemos llamar a `useState` y a ` useEffect` muchas veces en un momento y serán completamente independientes.
```

### Tip: Pasa información entre Hooks

Como los Hooks son funciones, podemos pasar información entre ellos.

> Para demostrar esto, vamos a usar otro componente de nuestro hipotético ejemplo de chat. Este es un recipiente de selector de mensaje de chat que muestra cuando el amigo seleccionado actualmente esta conectado:

Reviewer: ```suggestion
Para demostrar esto, vamos a usar otro componente de nuestro hipotético ejemplo de chat. Este un selector del destinatario del mensaje que muestra si el amigo seleccionado está conectado:
```

}
```

Mantenemos el amigo seleccionado actual en la variable de estado `recipientID`, y la actualizamos si el usuario elige a un amigo diferente en el selector `<select>`.

> Como la llamada al Hook `useState` nos da el ultimo valor de la variable de estado `recipientID`, podemos pasarla a nuestro Hook personalizado `useFriendStatus` como un argumento:

Reviewer: ```suggestion
Como la llamada al Hook `useState` nos da el último valor de la variable de estado `recipientID`, podemos pasarla a nuestro Hook personalizado `useFriendStatus` como un argumento:
```

```js
  const [recipientID, setRecipientID] = useState(1);
  const isRecipientOnline = useFriendStatus(recipientID);
```

> Esto nos permite saber cuando el amigo* actualmente seleccionado* esta en linea. Si elegimos un amigo diferente y actualizamos la variable de estado `recipientID`, nuestro Hook `useFriendStatus` va a desuscribirse de el amigo previamente seleccionado, y suscribirse al estado de el nuevo seleccionado.

Reviewer: ```suggestion
Esto nos permite saber cuando el amigo* actualmente seleccionado* esta en línea. Si elegimos un amigo diferente y actualizamos la variable de estado `recipientID`, nuestro Hook `useFriendStatus` eliminará su suscripción del amigo previamente seleccionado, y suscribirse al estado del nuevo seleccionado.
```

Reviewer 2: I believe it's `está` in "Está en linea".

And for the ending of the last sentence:
> y se suscribirá al estado del nuevo seleccionado.

## `usaTuImaginación()`

> Los Hooks personalizados ofrecen la flexibilidad de compartir la lógica que no era posible con los componentes de React antes. Puedes escribir Hooks personalizados que cubran una amplia gama de casos de uso como manejo de formularios, animación, suscripciones declarativas, temporizadores y probablemente muchos mas que no hemos considerado. Ademas, puedes construir Hooks que sean tan fáciles de usar como las características incorporadas de React.

Reviewer: ```suggestion
Los Hooks personalizados ofrecen la flexibilidad de compartir lógica que no era posible antes con los componentes de React. Puedes escribir Hooks personalizados que cubran una amplia gama de casos de uso como manejo de formularios, animación, suscripciones declarativas, temporizadores y probablemente muchos más que no hemos considerado. Ademas, puedes construir Hooks que sean tan fáciles de usar como las características incorporadas de React.
```

> Intenta resistirte a añadir la abstracción demasiado pronto. Ahora que los componentes funcionales pueden hacer mas, es probable que el promedio de componentes funcionales en tu base de código se amplíe. Esto es normal -- no te sientas como si *tuvieras* que dividirlo inmediatamente en Hooks. Pero también te animamos a empezar a descubrir casos donde un Hook personalizado podría ocultar la lógica compleja detrás de una interfaz simple, o ayudar a desenredar un componente desordenado.

Reviewer: ```suggestion
Intenta resistirte a añadir la abstracción demasiado pronto. Ahora que los componentes funcionales pueden hacer más, es probable que el promedio de componentes funcionales en tu base de código se amplíe. Esto es normal, no te sientas como si *tuvieras* que dividirlo inmediatamente en Hooks. Pero también te animamos a empezar a descubrir casos donde un Hook personalizado podría ocultar la lógica compleja detrás de una interfaz simple, o ayudar a desenredar un componente desordenado.
```

Reviewer 2: I'd remove the article `la` in `Intenta resistirte a añadir la abstracción`, I believe they refer to abstraction in general not an abstraction in particular.

> Por ejemplo, quizás tienes un componente complejo que contiene muchos estados locales que son gestionados en una forma ad-hoc. El `useState` no hace la centralización de actualización de lógica mas fácil, así que podrías preferir escribirlo como un reductor [Redux](https://redux.js.org/):

Reviewer: ```suggestion
Por ejemplo, quizás tienes un componente complejo que contiene muchos estados locales que son gestionados en una forma *ad-hoc*. El `useState` no hace la centralización de actualización de lógica mas fácil, así que podrías preferir escribirlo como un reductor [Redux](https://redux.js.org/):
```

Reviewer 2: You used `muchos estados locales`, that means `lots of local states`, in plural. But they say `a lot of local state`, which would become `mucho estado local` or `una gran cantidad de estado local`.

If we consider `useState` a name, then the article `El` on the second sentence shouldn't be necessary.

        text: action.text,
        completed: false
      }];
    // ... otras acciones ...
    default:
      return state;
  }
}
```

> Los reductores son muy convenientes para probar en aislamiento y escalar para expresar una lógica de actualización compleja. Puedes separarlos aun más en reductores mas pequeños si es necesario. Sin embargo, puedes también disfrutar los beneficios de usar el estado local de React, o puedes no querer instalar otra librería.

Reviewer: ```suggestion
Los reductores son muy convenientes para probar en aislamiento y escalar para expresar una lógica de actualización compleja. Puedes separarlos aun más en reductores mas pequeños si es necesario. Sin embargo, puedes también disfrutar los beneficios de usar el estado local de React, o puedes no querer instalar otra biblioteca.
```

Reviewer 2: At the beginning of the last sentence, I suggest this:
> Sin embargo, es posible que también te gusten los beneficios de ...

> Y si pudiéramos escribir un Hook `useReducer` que nos permita manejar el estado *local* de nuestro componente con un reductor? Una versión simplificada de esto podría verse así:

Reviewer: ```suggestion
¿Y si pudiéramos escribir un Hook `useReducer` que nos permita manejar el estado *local* de nuestro componente con un reductor? Una versión simplificada de esto podría verse así:
```

}
```

> La necesidad de gestionar estados locales con un reductor en un componente complejo es lo suficientemente común que hemos construido el Hook `useReducer` justo dentro de React. Vas a encontrarlo junto con otros Hooks incororados en la[referencia a la API de Hooks](/docs/hooks-reference.html).

Reviewer: ```suggestion
La necesidad de gestionar estados locales con un reductor en un componente complejo es lo suficientemente común que hemos construido el Hook `useReducer` justo dentro de React. Vas a encontrarlo junto con otros Hooks incorporados en la [referencia a la API de Hooks](/docs/hooks-reference.html).
```

Reviewer 2: Isn't `justo dentro de React` too much of a literal translation? 
Maybe leaving `dentro de React` is enough, leaving `justo` out.

Finally, for the last sentence, I suggest starting it this way:
> Lo encontrarás junto con otros Hooks ...
